proposals/2078-optimization-ol-facets.md

```diff
@@ -0,0 +1,78 @@
+# Proposal: Optimize query performance for OpenLineage facets
+
+Author: Willy Lulciuc ([@wslulciuc](https://github.com/wslulciuc))
+
+Created: 2022-08-18
+
+Discussion: [https://github.com/MarquezProject/marquez/issues/2078](https://github.com/MarquezProject/marquez/issues/2078)
+
+## Overview
+
+[OpenLineage](https://openlineage.io) was initially prototyped using Marquez, with the [initial draft](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md#010---2021-08-12) of the spec taking inspiration from Marquez's [data model](https://lucid.app/lucidchart/f918ce01-9eb4-4900-b266-49935da271b8/view?page=8xAE.zxyknLQ#). OpenLineage events are collected via [`POST` `/lineage`](https://marquezproject.github.io/marquez/openapi.html#tag/Lineage) calls, and can be queried via the `lineage_events` table using the `runID` associated with the event. The _current_ schema for the `lineage_events` table is defined below:
+
+### Table `lineage_events`
+
+| **COLUMN**    | **TYPE**     |
+|---------------|--------------|
+| run_uuid      | `UUID`       |
+| event_time    | `TIMESTAMPZ` |
+| event_type    | `TEXT`       |
+| event         | `JSONB`      |
+| job_name      | `TEXT`       |
+| job_namespace | `TEXT`       |
+| producer      | `TEXT`       |
+
+> **Note:** The table has an index on `run_uuid` and a _multicolumn_ index on `job_name`, `job_namespace`.
+
+The `event` column contains the raw OpenLineage event. When Marquez handles the event, metadata for datasets, jobs, and runs are inserted into their respective tables.
+
+OpenLineage's core model is extensible via _facets_. A `facet` is user-defined metadata and enables entity enrichment. Initially, returning dataset, job, and run facets via the REST API was not supported, but eventually added in release [`0.14.0`](https://github.com/MarquezProject/marquez/compare/0.13.1...0.14.0). The implementation was simple: when querying the `datasets`, `jobs`, or `runs` tables, also query the `lineage_events` table for facets.
+
+We knew the initial implementation would have to eventually be revisited. That is, OpenLineage events can easily exceed **>** **`10MBs`** resulting in out-of-memory (OOM) errors as facet queries require loading the raw `event` in memory, then filtering for relevant facets. This proposal outlines how we can optimize query performance for OpenLineage facets.
```

Comment:
10MB is not small for a single record, but it may not be big enough for resulting in OOM. Can you please elaborate the query example which causes an issue, which may load large number of lineage event records? Besides, we may want to address the high CPU usage problem by unnecessarily parsing big JSONB object when not needed, for example loading run facets when querying dataset.

Reply:
> 10MB is not small for a single record, but it may not be big enough for resulting in OOM

Very true, but if you load a lot of them, it will result in OOM:

```sql
          + "    LEFT JOIN LATERAL (\n"
          + "        SELECT run_uuid, event_time, event FROM lineage_events\n"
          + "        WHERE run_uuid = dv.run_uuid\n"
          + "    ) e ON e.run_uuid = dv.run_uuid\n"
          + "    UNION\n"
          + "    SELECT d.uuid, d.name, d.namespace_name, rim.run_uuid, lifecycle_state, event_time, event\n"
          + "    FROM selected_datasets d\n"
          + "    INNER JOIN dataset_versions dv ON dv.uuid = d.current_version_uuid\n"
          + "    LEFT JOIN runs_input_mapping rim ON dv.uuid = rim.dataset_version_uuid\n"
          + "    LEFT JOIN LATERAL (\n"
          + "        SELECT run_uuid, event_time, event FROM lineage_events\n"
          + "        WHERE run_uuid = rim.run_uuid\n"
          + "    ) e ON e.run_uuid = rim.run_uuid\n"
```

The partial query above is from [`DatasetDao.findDatasetByName()`](https://github.com/MarquezProject/marquez/blob/main/api/src/main/java/marquez/db/DatasetDao.java#L82). This specific query loads multiple OL events and is where we've seen the error. I should've added this context in the proposal (I'll do that now).